### PR TITLE
cosmos: fix partition-key equality/IN query panic (#4574)

### LINKED ISSUE #4574: Cosmos Rust SDK v0.35 still panics for full-container query with partition-key `IN` predicate
## Summary

`azure_data_cosmos` panics inside `azure_data_cosmos_driver` when running a full-container query filtered by the container partition key using `IN`.

This reproduces with a Cosmos DB SQL API container whose partition key path is `/id`.

## Versions

- `azure_data_cosmos`: `0.35.0`
- `azure_data_cosmos_driver`: `0.4.0`
- Rust: please fill in the output of `rustc --version`
- OS: Fedora Linux

## Container

- Azure Cosmos DB SQL API
- Partition key path: `/id`

## Minimal Repro

I created a standalone Rust binary repro that talks directly to
`azure_data_cosmos` and does not depend on Spin:

```bash
cd azure-cosmos-query-repro
export COSMOS_KEY="<redacted>"
RUST_BACKTRACE=full cargo run
```

The repro uses these non-secret defaults, overridable via environment variables:

```text
COSMOS_ACCOUNT="some-azure-cosmos-kv-store"
COSMOS_DATABASE="some-azure-cosmos-kv-store-db"
COSMOS_CONTAINER="some-azure-cosmos-kv-store-container"
COSMOS_REGION="East US"
```

The app:

1. Upserts two documents into a `/id`-partitioned container:

```json
{ "id": "repro:many1", "value": [...] }
{ "id": "repro:many2", "value": [...] }
```

2. Confirms point reads work:

```rust
client.read_item(key.clone(), key, None).await?;
```

3. Confirms a simple full-container query works:

```rust
client
    .query_items::<Pair>(
        Query::from("SELECT * FROM c"),
        FeedScope::full_container(),
        None,
    )
    .await?;
```

4. Runs the failing full-container query filtered by the partition key:

```rust
let query = Query::from("SELECT * FROM c WHERE c.id IN (@key0, @key1)")
    .with_parameter("@key0", "repro:many1")?
    .with_parameter("@key1", "repro:many2")?;

client
    .query_items::<Pair>(query, FeedScope::full_container(), None)
    .await?;
```

## Expected Behavior

The SDK should return query results or a structured error.

## Actual Behavior

The SDK panics:

```text
thread 'tokio-rt-worker' panicked at .../azure_data_cosmos_driver-0.4.0/src/driver/dataflow/planner.rs:258:83:
topology provider must return ranges that overlap the query plan EPK range
```

Relevant stack frames:

```text
azure_data_cosmos_driver::driver::dataflow::planner::build_sequential_drain
azure_data_cosmos_driver::driver::cosmos_driver::CosmosDriver::plan_operation
azure_data_cosmos::clients::container_client::ContainerClient::query_items
azure_cosmos_query_repro::partition_key_in_query
```

## Notes

This looks related to the partition-key-range/query-planner area touched by Azure/azure-sdk-for-rust#4467, but the repro is not a split/continuation case.

The query is a full-container query whose predicate is `c.id IN (...)`, and `id` is also the container partition key. A possible explanation is that the gateway query plan returns narrow or point-like EPK ranges for the partition-key `IN` predicate, and the planner later treats those ranges as normal `[min, max)` ranges before intersecting them with topology.

In any case, the SDK should not panic from this input.

### PR DESCRIPTION
**Pull request by ananth7592**
## Problem

A cross-partition query that filters on the partition key with an **equality or `IN` predicate** (e.g. `SELECT * FROM c WHERE c.pk = @pk`, or the original repro `WHERE c.id IN (@a, @b)`) panicked the query worker and hung the caller indefinitely:

```
thread 'tokio-rt-worker' panicked at .../planner.rs:
topology provider must return ranges that overlap the query plan EPK range
```

The partition key is in the SQL text (not supplied as a request-option partition key), so this is the predicate-routing path, distinct from the already-working `FeedScope::Partition` path.

## Root cause

The gateway query plan returns a **closed point** EPK range `[X, X]` (`isMinInclusive == isMaxInclusive == true`) for a partition-key equality (`IN` is just N such points). `FeedRange` is half-open `[min, max)`, so the point became `[X, X)` — the empty set. `intersect_feed_ranges` requires `min < max`, returned `None`, and the `.expect()` panicked. Because the panic is on the worker thread, no completion is posted and the caller hangs.

## Fix

Normalize the closed point `[X, X]` to the half-open **EPK-range window** `[X, successor(X))` and route it like any other range: `partitionkeyrangeid` **plus** a `start`/`end-epk` pair, with `x-ms-read-key-type: EffectivePartitionKeyRange`. The window is honored by the gateway and returns exactly the matching document(s); the SQL predicate still filters to the exact value(s).

This is the **"always normalize to min-inclusive / max-exclusive"** model requested by the service team and used by the .NET SDK — it keeps every EPK comparison a plain half-open range check and avoids special-casing points in the overlap logic. The successor is computed at the container's **full HPK width** (trailing-zero padding), matching .NET's HPK normalization, so the window is the tightest correct one the backend expects.

### Salient code changes

```
query_range_to_feed_range(query_range, normalized_len):
  # transform ONLY a closed point (equality / IN value); pass everything else through
  if is_max_inclusive && min == max:                 # point [X, X]
      max = match normalized_len:
          Some(len) => X.normalized_successor(len)    # pad to full HPK width, then +1
          None      => X.successor()                  # V1: width-preserving +1
  # closed NON-point ranges (e.g. HPK prefix bounds) and half-open ranges: unchanged
```

- **`EffectivePartitionKey`** (`effective_partition_key.rs`) — new `successor()` (big-endian +1 with carry, width-preserving) and `normalized_successor(len)` (zero-extend to full width, then `successor`); `normalized_epk_len(def)` = `16 * paths().len()` for V2 Hash/MultiHash, `None` for V1.
- **`query_range_to_feed_range`** (`planner.rs`) — turns a closed **point** into `[X, successor(X))` (full-width for known EPK width, plain successor for V1). A closed **non-point** range — e.g. an HPK **prefix** bound, already at partition-boundary granularity — is passed through unchanged; advancing it would over-extend the prefix band and drop owning partitions.
- **`plan_fresh` / `plan_resume_from_saved_snapshot`** — each normalized point resolves to its owning partition(s) and emits the narrow `[X, successor(X))` window as a `start`/`end-epk` pair alongside `partitionkeyrangeid`. Co-located `IN` values emit one window each within their shared partition; the resume path attaches the covering continuation token and drops points below the cursor.
- **`PartitionKeyRangeCache` / `topology`** — a range whose lower bound lands exactly on a partition boundary resolves via the boundary-correct point lookup, so a point maps to the correct owning partition.
- The planner's topology-intersection path now returns a `CosmosError` instead of a worker-thread panic for any future unservable plan.

## Behavior change

Equality / `IN` predicate queries now route as narrow per-value EPK windows instead of panicking. Results are correct and de-duplicated (the SQL predicate does the value filtering). The complete-PK `FeedScope::Partition` path (logical-PK header) is unchanged. No wire behavior changes for non-point ranges.

## Testing & coverage

- Unit coverage verifies the core routing behavior: equality / `IN` points are normalized to narrow `[X, successor(X))` windows, full-width HPK padding is applied when needed, and non-point ranges such as HPK prefix bounds remain unchanged.
- Emulator coverage exercises both regular partition-key equality / `IN` queries and HPK full-key equality across partition boundaries, while keeping the existing HPK prefix/topology scenarios green.
- Validation is green in the driver unit tests, the full in-memory emulator suite, and `cargo clippy --all-features`.

**Live account validation**

Validated against a real account using:
- a single-hash `/pk` container
- a 3-level HPK container (`/state,/city,/county`) physically split at `state=WA` and `city=city-183`

Results:
- `/pk` equality and `IN` queries returned exactly the expected documents, including mixed present / missing values.
- HPK full-key equality returned exactly one matching document on either side of the physical split, and no rows for a missing key.
- The HPK prefix query `WHERE c.state = 'WA'` returned all `WA` documents, confirming that non-point prefix ranges still route correctly.

Fixes #4574.

### CODE REVIEW

File: sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md [hunk 1]
```diff
@@ -46,6 +46,7 @@
...
 - Fixed per-attempt `activity_id` in diagnostics serializing as `null` on transport failures; it is now seeded from the operation-level activity ID. ([#4602](https://github.com/Azure/azure-sdk-for-rust/pull/4602))
 - Fixed `system_usage.cpu` in diagnostics emitting the literal string `"empty"`; it now serializes as a structured `{ samples, status }` object when no CPU samples are available. ([#4602](https://github.com/Azure/azure-sdk-for-rust/pull/4602))
+- Fixed a worker-thread panic when a cross-partition query filters on the partition key with an equality / `IN` predicate (e.g. `WHERE c.id IN (@a, @b)` on an `/id`-partitioned container). The gateway returns a closed point EPK range `[X, X]` for such predicates; the planner now represents it as a distinct effective-partition-key point that routes to its owning physical partition (filtered by the SQL predicate) instead of collapsing to an empty `[X, X)` range that panicked the query worker and deadlocked the caller. Co-located `IN` values are de-duplicated so they no longer return duplicate documents. ([#4574](https://github.com/Azure/azure-sdk-for-rust/issues/4574))
```
[the remark this answers is not in the document]
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md]: Done in 0e0ef9a6 — the entry now links to /pull/4638 (the Fixes #4574 reference is kept in the PR description).

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs [hunk 2]
```diff
@@ -208,21 +209,47 @@ async fn plan_fresh(
...
+                    intersect_feed_ranges(&resolved_range.range, &feed_range).ok_or_else(|| {
+                        topology_range_not_overlapping_error(&resolved_range.range, &feed_range)
+                    })?;
+                let target = RequestTarget::effective_partition_key_range(
+                    range,
+                    resolved_range.partition_key_range_id,
+                    resolved_range.range,
+                );
+                nodes.push(Box::new(Request::new(Arc::clone(operation), target, None)));
+            }
```
[the remark this answers is not in the document]
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs]: Done in 0e0ef9a6 — plan_fresh now computes 
ange in the if/else and shares a single RequestTarget/Request construction, matching plan_resume_from_saved_snapshot. No behavior change; the point-routing and dedup tests still pass.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs [hunk 3]
```diff
@@ -256,18 +283,33 @@ async fn plan_resume_from_saved_snapshot(
...
+            // A point (equality / `IN`) targets its whole owning partition; a
+            // half-open range is clipped to the query range. Computing the
+            // point's intersection would collapse `[X, X)` to empty (issue
+            // #4574), so the whole resolved partition is used instead.
+            let leaf_scope = if is_point {
+                if !point_targeted_partitions.insert(resolved_range.partition_key_range_id.clone())
+                {
+                    continue;
+                }
+                resolved_range.range.clone()
```
[the remark this answers is not in the document]
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs]: Done in 0e0ef9a6 — added two resume-path regression tests: 
esume_in_predicate_drops_point_partition_below_cursor (cursor clipping drops a drained point partition) and 
esume_in_predicate_colocated_points_collapse_and_carry_continuation (co-located IN points de-duplicate to one leaf while carrying the saved server continuation). Both confirm the resume path neither drops nor duplicates a partition.

File: sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs [hunk 4]
```diff
@@ -48,6 +48,21 @@ enum FeedRangeRepr {
...
+    /// `IN` predicate on the partition key (`WHERE c.pk = @pk`): a closed point
+    /// range `[X, X]`. Unlike [`FeedRangeRepr::LogicalPartition`] it carries no
+    /// [`PartitionKey`] (the gateway only hands back the *hashed* EPK, which
+    /// cannot be reversed), so it routes by resolving the owning physical
+    /// partition rather than by emitting a logical-partition-key header.
+    ///
+    /// Representing the point explicitly — rather than as a degenerate
+    /// half-open `Range { min == max }` — keeps it from collapsing to the empty
+    /// set in `[min, max)` interval math (see issue #4574).
+    EpkPoint(EffectivePartitionKey),
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs]: No need to introduce yet another represenation - just convert the point EPK to a normalized EPK (minInclusive==true, maxInclusive==False) - can be done by settng max to pointEPK + 1

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Fix should just force normalization to EPK (MinInclusive and MaxExclusive)

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs [hunk 5]
```diff
@@ -208,15 +209,39 @@ async fn plan_fresh(
...
-            );
+            // A point (equality / `IN`) targets its *whole* owning partition by
+            // partition-key-range id (no `x-ms-start-epk` / `x-ms-end-epk`); the
+            // SQL predicate filters within the partition, mirroring .NET's
+            // "EpkRange spans exactly one physical partition" routing. A
+            // half-open range is clipped to the query range as before.
+            let range = if is_point {
+                // De-duplicate co-located points: a second `IN` value hashing
+                // into an already-targeted partition is fully redundant.
+                if !point_targeted_partitions.insert(resolved_range.partition_key_range_id.clone())
```
Comment (NaluTripician) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs]: Dedup is only consulted on the point branch — half-open `query_range`s never insert into or check `point_targeted_partitions`. The comment's justification holds for IN/equality-only plans, but if the gateway can ever return one plan with both a closed point and an overlapping half-open range that resolve to the same partition `P` (e.g. `WHERE c.pk = @x OR c.pk > @y`), this emits a whole-partition request for `P` (no epk headers) **and** a clipped sub-range request for `P` (with start/end-epk), both carrying the same full predicate → duplicate documents.

Could you confirm whether the gateway can emit that mixed shape? If yes, the dedup needs to be aware of half-open coverage too; if it's provably unreachable, a one-line assertion/comment documenting that invariant would prevent a future gateway change from silently reintroducing duplication.

**Review by NaluTripician — COMMENTED** [review]
Reviewed this end-to-end (traced the change through `feed_range.rs`, `planner.rs`, `topology.rs`, the routing-map cache, and `effective_partition_key.rs`). Nice fix — giving the bare hashed point its own `EpkPoint` variant instead of overloading a degenerate `[X, X)` range is the right call, the containment routing matches .NET, and converting the worker-thread `.expect()` into a propagating `CosmosError` genuinely unblocks the caller (it rides the same `?` path that `validate_query_plan` already uses, so it reaches the caller rather than dying on the worker thread).

No blocking issues. A few things I confirmed are safe so they don't get re-litigated:
- **Hash V1/V2** is a non-issue here — the point EPK is the gateway's already-hashed hex string, compared lexically against partition bounds in the same space; this path never re-hashes.
- **Boundary resolution** uses `min <= X < max` consistently with `resolve_overlapping_ranges` and `overlaps`; a point at a partition `min` resolves to the upper partition everywhere.
- **No remaining worker-thread panics** in the planner routing path — both `.expect()` sites are gone and the point branch short-circuits before `intersect_feed_ranges`.
- **Resume path** only ever operates on the resolved half-open partition range, so `EpkPoint` never leaks into cursor/coverage math.

Three things worth addressing, none blocking:

**1. (please confirm before merge) Dedup only guards the point branch.** Half-open ranges never consult `point_targeted_partitions`. If the gateway can ever emit a single plan containing both a closed point and an overlapping half-open range that resolve to the same physical partition (e.g. `WHERE c.pk = @x OR c.pk > @y`), the point emits a whole-partition request and the half-open emits a clipped sub-range request — both carrying the full SQL predicate — and the same documents come back twice. Inline comment on the exact spot. If that mixed shape is provably unreachable, a one-line invariant assertion/comment would lock it in (same defensiveness you already added for the degenerate-range case).

**2. (minor, latent) `EpkPoint` doesn't round-trip through serialization.** `to_json` serializes via `min_inclusive()/max_exclusive()`, so a point becomes `{min:X, max:X}`, and `from_json` rebuilds it as a degenerate `Range{X,X}` (not an `EpkPoint`) — the original empty-set shape. It's safe today only because the planner serializes the resolved partition's half-open range, never the `EpkPoint` itself. A `debug_assert!`/comment on `to_json` guarding against serializing a point would stop a future `to_string()` from silently corrupting it.

**3. (minor) `is_subset_of` wasn't made point-aware** the way `overlaps` was — latent inconsistency (it'd treat the exclusive upper bound as contained for a point). No production caller passes a point today, so just future-proofing.

Test-wise the added cases genuinely prove the headline claims (no-epk-header whole-partition targeting, co-located dedup, cross-partition one-each, the `min`-boundary resolution the old code missed, the logical-PK regression, both resume variants). Gaps that'd be nice to add: empty `IN` list, `IN` across 3+ partitions, and — if reachable — the mixed point/half-open plan from #1.

File: sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs (hunk 4 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs]: Done — thanks, this is exactly the direction we took in `4dcf8bae8`. The `EpkPoint` representation (and its accessors, point-aware `overlaps`, the topology containment branch, and the `resolve_range_containing` cache helper) is removed. `query_range_to_feed_range` now normalizes the gateway's closed point `[X, X]` to the half-open `[X, successor(X))` (`isMinInclusive=true, isMaxInclusive=false`), where `successor()` is a big-endian byte increment of the EPK — i.e. `max = pointEPK + 1`, as you suggested. `[min, max)` is now the single universal representation, so the point flows through the normal routing / intersection paths with no special-casing. I cross-checked against the Java SDK's `FeedRangeInternal.normalizeRange` → `addToEffectivePartitionKey(max, +1)`, which does the same byte increment.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs (hunk 5 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs]: Good catch — and the normalization in `4dcf8bae8` removes the asymmetry that caused this entirely, so the concern no longer applies. There's no longer a point branch, no `point_targeted_partitions` dedup, and no whole-partition-without-epk-headers targeting. A closed point is normalized to `[X, successor(X))` and clipped via `intersect_feed_ranges` exactly like any half-open range, so it's emitted as an EPK-scoped sub-range request (`start/end-epk`), not a whole-partition request. In your `pk = @x OR pk > @y` example, the point now contributes a request bounded to `[X, X+1)` rather than the whole partition, so it can't blanket-overlap the half-open request's range.

What remains is the general property that applies to *any* plan (all-half-open included): if the gateway ever returned overlapping `query_ranges`, the overlap would be double-scanned. Query-plan ranges are disjoint by contract, so we rely on that the same way the half-open path always has — there's no longer a point-specific way to reintroduce duplication.

File: sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs (hunk 4 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/models/feed_range.rs]: Closing the loop: this is now validated end-to-end against a live account (standard Gateway, HTTP/1.1, Gateway 2.0 off). The closed point [X, X] is normalized to [X, successor(X)) as you suggested, with the successor computed at the container's full HPK width (trailing-zero padding, matching .NET's HPK normalization). A 3-level HPK container (/state,/city,/county) physically split at state=WA & city=city-183 returns exactly the matching document for full-key equality on either side of the split, and a partial-key WHERE c.state='WA' prefix (a non-point range, left un-normalized) returns all WA docs. Thanks again for steering us to the single normalized representation.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs (hunk 5 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs]: Closing the loop, including the exact mixed-shape risk you flagged: validated end-to-end on a live account. The final code normalizes only closed points (min == max) to [X, successor(X)) and passes every non-point range (e.g. HPK prefix bounds) through unchanged, so points and half-open ranges are both clipped via intersect_feed_ranges identically -- there's no whole-partition-without-EPK-headers path left, so a pk = @x OR pk > @y plan can't produce the duplicate you described. Live HPK results: full-key equality on each side of a real split returns exactly one document; WHERE c.state='WA' returns all WA docs (confirming the non-point prefix range neither under- nor over-covers). The remaining reliance is just the general query-plan disjointness contract -- the same one the all-half-open path has always depended on.

File: sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/partition_key_equality.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    let container = client
+        .database_client(&db_name)
+        .container_client("docs")
+        .await
+        .unwrap();
+
+    for item in dataset() {
+        container
+            .create_item(PartitionKey::from(item.pk.clone()), &item.id, &item, None)
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/partition_key_equality.rs:109]: NIT: Please use the new test helper create_seed_item - it helps to avoid flaky tests due to some transient error when seeding test items. (ok, to do in subsequent PR after the release)

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs [hunk 7]
```diff
@@ -586,11 +610,60 @@ async fn plan_resume_from_saved_snapshot(
 }
 
 /// Converts a query-plan EPK range to a [`FeedRange`].
+///
+/// The gateway returns a *closed* range `[a, b]`
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs:614]: NIT: My preference would be to have a few really simple unit tests only testing this algorithm. Basically testing the function for normalization, passing a few EPKs (some with shorter than normalized length) and PK Definition as well as expetced value to encode this algorithm in a simple test.

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs:614]: The tests below test this implicitly - but all in combination with queries - which is good to have - but this normalization can easily be unit-tested and makes it easier to watch our for edge-cases that might be msising in test coverage.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs [hunk 8]
```diff
@@ -371,7 +371,12 @@ pub(crate) enum SortOrder {
 /// [`RawQueryPlan::resolve`] to hash structured PK values into proper EPK hex.
 #[derive(Debug, Deserialize, Serialize, PartialEq)]
 #[serde(rename_all = "camelCase")]
-#[allow(dead_code)] // Inclusivity flags are wire-format; planner treats ranges uniformly.
+// `is_min_inclusive` is wire-format only: query-plan ranges are always
+// min-inclusive and `FeedRange` is half-open `[min, max)`. `is_max_inclusive`
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs]: Why even keep the dead code here? Just remove it?

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs (hunk 7 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs:614]: effective_partition_key.rs has some unit test goign in this direction - they do not cover whether we correctly normalize the length - but they have overflow tests. So, I think overall the test coverage is suffiicent.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM except for a few  NITs.

File: sdk/cosmos/azure_data_cosmos_driver/src/models/effective_partition_key.rs [hunk 9]
```diff
@@ -52,6 +52,77 @@ impl EffectivePartitionKey {
...
+    /// Normalizing to `normalized_len_bytes` first makes `[X, successor(X))`
+    /// the tightest correct half-open range. This mirrors the .NET SDK, which
+    /// always zero-extends hierarchical EPKs to full length before comparing.
+    ///
+    /// `normalized_len_bytes` is the full EPK width in bytes for the container's
+    /// partition key definition — `16 * paths().len()` for V2 hash / MultiHash
+    /// (see [`normalized_epk_len`]). When `self` is already at least that wide
+    /// (the common full-key case), no padding is applied and the result equals
+    /// [`successor`](Self::successor).
+    pub(crate) fn normalized_successor(
```
Comment (analogrelay) [inline sdk/cosmos/azure_data_cosmos_driver/src/models/effective_partition_key.rs:87]: This makes me wonder if we should just be storing these as normalized length. So all `EffectivePartitionKey` instances are exactly the same length, we just pad out during construction and "unpad" (remove unnecessary trailing `0`s) for rendering the hex string. That would also solve the comparison quirks, right?

Comment (analogrelay) [reply sdk/cosmos/azure_data_cosmos_driver/src/models/effective_partition_key.rs:87]: Ah, v1 probably complicates that. But we _could_ store the version inside the EPK to make it clear. As we've previously discussed, there's not really a difference (at least for the things we care about) between Hash/MultiHash, it's all just the same v2 encoding format, just with a variable number of allowed components.

**Review by analogrelay — APPROVED** [review]
Nothing blocking, all of it can be handled in follow-up if we want.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs [hunk 10]
```diff
@@ -386,11 +379,10 @@ pub(crate) struct QueryRange {
     #[serde(deserialize_with = "string_or_json")]
     pub max: String,
 
-    /// Whether the minimum value is inclusive.
-    #[serde(deserialize_with = "bool_from_int_or_bool")]
-    pub is_min_inclusive: bool,
-
-    /// Whether the maximum value is inclusive.
```
Comment (analogrelay) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs:388]: Non-Blocking, but I actually don't know if I agree with getting rid of this. If the gateway ever sent us a query plan with `isMinInclusive: false`, we would silently ignore that and perhaps execute the query incorrectly. It's non-blocking because I know the chance of that actually happening is _extremely_ unlikely, but I still prefer asserting our expectation here (returning some internal error to the user if that ever happens) rather than silently ignoring an incorrect range.

I definitely agree with removing it if we don't use it _at all_ (hence with removing the dead_code suppression) but I actually would prefer a quick sanity check for `is_min_inclusive == true` as soon as we get back a query plan from the Gateway (or native library, in the future).

**Review by analogrelay — APPROVED** [review]

File: sdk/cosmos/azure_data_cosmos_driver/src/models/effective_partition_key.rs (hunk 9 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/models/effective_partition_key.rs:87]: I don't think Storing makes sense especially because it complicates cases like FeedRanges and not required in most cases.

File: sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/partition_key_equality.rs (hunk 6 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/partition_key_equality.rs:109]: Will do in a subsequent PR for this, That;s good idea to keep the CI runs clean.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs (hunk 8 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs]: Removed it along with the allow(dead_code). Kept a guard instead of just dropping the flag - resolve now rejects isMinInclusive=false with an internal error, so a bad plan fails loud rather than silently mis-routing. 148be05c4

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs (hunk 7 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/planner.rs:614]: Added a direct unit test - normalized_successor_encodes_the_planner_algorithm_end_to_end - wiring normalized_epk_len(def) into normalized_successor for single-hash and 3-level HPK. Length normalization is covered on its own now, not just implicitly via the query tests. d20c0b6fe

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs (hunk 10 above)
Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs:388]: Went with your suggestion - resolve now rejects isMinInclusive=false with an internal error rather than ignoring it, so we fail loud instead of mis-routing. Flag stays off QueryRange (a missing wire value defaults to min-inclusive). Added a negative test. 148be05c4

Comment (ananth7592) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/dataflow/query_plan.rs:388]: On the other hand, I am thinking since this is non blocking might give it a shot as a follow up PR, Let me rever this

**Review by analogrelay — APPROVED** [review]
